File: ruby/test/test_datasources.rb
# frozen_string_literal: true

require_relative './base_template'

SAMPLE_DATASOURCE_PARAMS = {
  description: 'Ruby Integration Test Data Source',
  title: 'Ruby Integration Test Data Source',
  source: 'api',
  config: '{"type":"json","method":"GET","transform":"products","url":"https://dummyjson.com/products","headers":"{}"}'
}.freeze

# All tests related to Onvo's datasource endpoints
class DatasourcesTest < BaseTemplate
  def before_all
    super
    @sample_datasource_id = @onvo.datasources.create(SAMPLE_DATASOURCE_PARAMS)['id'] # skipping test create
  end

  def test_list_datasources
    assert_silent { @onvo.datasources.list }
  end

  # TODO : diagnose and fix 500 error
  # def test_upload_file
    # assert_silent do
    #   @onvo.datasources.upload_file(@sample_datasource_id, 'test/fixtures/test.json')
    # end
  # end

  def test_initialize_datasource
    assert_silent do
      @onvo.datasources.initialize_datasource(@sample_datasource_id)
    end
  end

  def test_get_datasource
    assert_silent { @onvo.datasources.get(@sample_datasource_id) }
  end

  def test_update_datasource
    assert_silent do
      @onvo.datasources.update(@sample_datasource_id, { title: 'Renaming Test' })
    end
  end

  def after_all
    @onvo.datasources.delete(@sample_datasource_id)
  end
end

File: python/tests/test_datasources.py
from .base_test import BaseTest
import os

SAMPLE_DATASOURCE_PARAMS = {
    "description": "Python Integration Test Data Source. Delete if seen.",
    "title": "Python Integration Test Data Source",
    "source": "api",
    "config": '{"type":"json","method":"GET","transform":"products","url":"https://dummyjson.com/products","headers":"{}"}',
}


class TestDatasources(BaseTest):

    @classmethod
    def setUpClass(cls):
        super().setUpClass()
        cls.sampleDatasourceId = cls.onvoSDK.datasources.create(
            SAMPLE_DATASOURCE_PARAMS
        )[
            "id"
        ]  # Skipping test create

    def test_list(self):
        self.assertShouldRaise(None, self.onvoSDK.datasources.list)

    def test_get(self):
        self.assertShouldRaise(
            None, lambda: self.onvoSDK.datasources.get(self.sampleDatasourceId)
        )

    def test_initialize(self):
        self.assertShouldRaise(
            None, lambda: self.onvoSDK.datasources.initialize(self.sampleDatasourceId)
        )

    def test_update(self):
        self.assertShouldRaise(
            None,
            lambda: self.onvoSDK.datasources.update(
                self.sampleDatasourceId, {"title": "Renaming Test"}
            ),
        )

    # TODO: diagnose and fix 500: Internal server error
    # def test_upload_file(self):
    #     test_file_path = os.path.join(os.path.dirname(__file__), "python_onvo_test.csv")
    #     with open(test_file_path, "ab") as f:
    #         f.write(b"a,b,c\n1,2,3\n4,5,6\n")
    #     self.assertShouldRaise(
    #         None,
    #         lambda: self.onvoSDK.datasources.upload_file(
    #             self.sampleDatasourceId, test_file_path
    #         ),
    #     )
    #     os.remove(test_file_path)

    @classmethod
    def tearDownClass(cls) -> None:
        cls.onvoSDK.datasources.delete(cls.sampleDatasourceId)  # Skipping test delete
        super().tearDownClass()
